### PR TITLE
Optional recompute-in-backward for the v5 grouped MoE path

### PR DESCRIPTION
**Pull request by danielhanchen**
### Problem

The native grouped MoE forward (`forward_native_grouped_mm`) feeds the dequantized bf16 expert weight stack to `torch._grouped_mm`. Native autograd then keeps that dense stack (`[E, hidden, 2*inter]` and `[E, inter, hidden]`) alive for the whole backward so it can form `dX = grad @ W^T`. The base expert weight is frozen and LoRA is applied as a separate additive grouped GEMM, so the base matmul only needs its weight for `dX`, never a base `dW`. That makes the pinned bf16 stack pure overhead that can be recomputed.

### Change

Add a small custom autograd Function that saves only `(inputs, offsets)` plus a weight provider, and rebuilds the frozen base weight from the 4-bit `Params4bit` in backward:

- `_GroupedMMRecompute` + `_base_grouped_mm` + `_moe_recompute_enabled` in `moe_utils.py`, wired into the gate_up and down base matmuls of `forward_native_grouped_mm`.
- A matching branch in `moe_utils_bnb4bit.py` so the no-LoRA bnb-4bit path defers its dequant into the provider instead of pre-dequantizing and pinning the stack.

One shared change covers every model that routes through the grouped backend (`qwen3_moe`, `qwen3_5_moe`, `gemma4`, `glm4_moe_lite`, and the others).

### Opt-in and safety

- Off by default (`UNSLOTH_MOE_RECOMPUTE=1` to enable). When off, `_base_grouped_mm` is exactly the prior eager call, so the path is byte-for-byte unchanged.
- Enabled only on the separated-LoRA path, with a frozen base (bnb-4bit with quant_state, or plain-frozen), and `torch._grouped_mm` support. So merged-LoRA, full finetuning, the fp8 path, the `w1`/`w3`/`w2` branch, and non-CUDA are untouched.
- The recompute output and `dX` are bit-identical to the eager path (verified on synthetic stacked experts, including zero-token experts).

### Results

4-bit QLoRA, 1x B200, seq 2048, gradient checkpointing on, 16 steps (off vs on):

| model | peak (off) | peak (on) | tok/s off | tok/s on |
|-------|-----------|-----------|-----------|----------|
| Qwen3-30B-A3B-Instruct-2507 | 26.27 GiB | 24.99 GiB | 2678 | 2557 |
| Qwen3.6-35B-A3B | 33.89 GiB | 32.36 GiB | 702 | 703 |
| gemma-4-26B-A4B | 24.21 GiB | 22.70 GiB | 2872 | 2759 |
| GLM-4.7-Flash | 23.46 GiB | 22.40 GiB | 2317 | 2246 |

About 1.0 to 1.5 GiB (4 to 6 percent) lower peak training memory at a small speed cost, with per-step loss matching. It is opt-in because gradient checkpointing already bounds most of the per-layer dequant, so the win is modest but free for memory-constrained or longer-context runs.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
### Benchmark: adaptive gradient-checkpoint-aware recompute (before vs after)

Quantified the adaptive `_moe_recompute_default` policy on the real grouped 4-bit expert path.

**Setup:** `unsloth/gpt-oss-20b-unsloth-bnb-4bit`, attention-only LoRA, bf16, `per_device_train_batch_size=2` x `gradient_accumulation_steps=3`, `seq=2048`, 61 steps, single B200. The grouped path engaged in every run (17568 `_base_grouped_mm` calls each). "Before" is the two static choices via `UNSLOTH_MOE_RECOMPUTE` (`=1` always recompute, `=0` always pin); "after" is the adaptive default (unset). Each regime's trio was run on one GPU for clean deltas, with an adaptive-vs-itself replicate to establish the run-to-run noise floor (**0.5%**). Metrics are the mean over post-warmup steps; peak GB is `max_memory_allocated` across steps.

**Gradient checkpointing ON (the default)**

| Variant | Peak GB | Step ms | tok/s |
|---------|---------|---------|-------|
| always-recompute (`=1`, before) | 16.40 | 1908 | 6439 |
| **adaptive (unset, after)** | 16.88 | **1829** | **6718** |
| adaptive replicate | 16.88 | 1839 | 6682 |
| always-pin (`=0`) | 16.88 | 1825 | 6734 |

**Gradient checkpointing OFF**

| Variant | Peak GB | Step ms | tok/s |
|---------|---------|---------|-------|
| always-recompute (`=1`) | 16.74 | 2691 | 4566 |
| **adaptive (unset, after)** | 16.74 | 2686 | 4574 |
| adaptive replicate | 16.74 | 2672 | 4598 |
| always-pin (`=0`, before) | 17.23 | 2477 | 4961 |

**Deltas (both well above the 0.5% noise floor):**

- With gradient checkpointing: adaptive is **+4.1% faster** than always-recompute (6718 vs 6439 tok/s) and matches always-pin (within 0.2%), at a cost of **+0.48 GB** (the momentary pinned stack).
- Without gradient checkpointing: adaptive stays on the lean path, **16.74 GB, identical to always-recompute**, and avoids always-pin's **+0.49 GB**.

So the adaptive default is never worse than the better static choice in each regime: it takes pin's speed when the pin is momentary (inside a checkpoint recompute pass) and recompute's memory safety otherwise, with no flag needed.

**Mechanism** (probe of the recompute decision at each grouped call):

- GC on, adaptive: the forward pass returns `recompute=True` (moot, it runs under `no_grad`), and the backward re-run returns `recompute=False` (pins). Split was exactly 50/50.
- GC off, adaptive: every call returns `recompute=True` (recompute).

**Loss is numerically transparent.** The cross-variant final-loss spread is within the identical-code noise floor (adaptive vs its own replicate: `max_loss_diff` 0.0144 with GC, 0.245 without), so the spread is bf16 `_grouped_mm` atomic nondeterminism, not the recompute choice. `UNSLOTH_MOE_RECOMPUTE` remains a manual override (`1` force recompute, `0` force pin).

File: unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py [hunk 1]
```diff
@@ -141,6 +141,17 @@ def forward_moe_backend_bnb4bit(self, hidden_states, top_k_index, top_k_weights)
...
+    # Fix 3: defer dequant into forward_native_grouped_mm's providers (recomputed in
+    # backward) instead of pre-dequantizing the full bf16 stack and pinning it.
+    if (
+        os.environ.get("UNSLOTH_MOE_RECOMPUTE", "0") == "1"
+        and select_moe_backend() == "grouped_mm"
+        and _is_bnb4bit_param(self.gate_up_proj)
+        and _is_bnb4bit_param(self.down_proj)
+    ):
+        _log_moe_bnb4bit_backend_once(self, "Unsloth: MoE bnb4bit grouped_mm with backward-recompute.")
+        return forward_native_grouped_mm(self, hidden_states.to(target_dtype), top_k_index, top_k_weights)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py:160]: _get_base_weight now restores _original_shape and casts to the input dtype for Params4bit, so the grouped-mm recompute provider gets the logical 3D expert stack. Fixed.

File: unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py [hunk 2]
```diff
@@ -141,6 +141,17 @@ def forward_moe_backend_bnb4bit(self, hidden_states, top_k_index, top_k_weights)
     if not target_dtype.is_floating_point:
         target_dtype = torch.bfloat16
 
+    # Fix 3: defer dequant into forward_native_grouped_mm's providers (recomputed in
+    # backward) instead of pre-dequantizing the full bf16 stack and pinning it.
+    if (
+        os.environ.get("UNSLOTH_MOE_RECOMPUTE", "0") == "1"
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py]: Done. The bnb4bit grouped_mm dispatch now gates on _moe_recompute_default() like the plain path, so it keeps the packed 4-bit weights and only pre-dequantizes when the policy says pin. That also removes the case where a pre-dequantized dense stack was re-held for a backward recompute.

File: unsloth_zoo/temporary_patches/moe_utils.py [hunk 3]
```diff
@@ -190,6 +292,51 @@ def _check_torch_grouped_mm_supported():
...
+# Some CUDA builds silently miscompute torch._grouped_mm for a transposed bf16 view preceded by a
+# broadcast op (pytorch/pytorch#186365, Blackwell + torch 2.11/2.13). This probe checks the view
+# matches the contiguous copy so _grouped_mm_with_backward_fix can skip the copy only when safe.
+_TRANSPOSED_VIEW_GROUPED_MM_SAFE = None
+
+
+def _transposed_view_grouped_mm_is_safe():
+    global _TRANSPOSED_VIEW_GROUPED_MM_SAFE
+    if _TRANSPOSED_VIEW_GROUPED_MM_SAFE is not None:
+        return _TRANSPOSED_VIEW_GROUPED_MM_SAFE
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils.py:334]: Keeping a single cached bool for now, mirroring _TORCH_GROUPED_MM_SUPPORTED; we assume a homogeneous GPU set within one process.

File: unsloth_zoo/temporary_patches/moe_utils.py [hunk 4]
```diff
@@ -190,6 +272,52 @@ def _check_torch_grouped_mm_supported():
...
+                    ok = False; break
+                if ref is None:
+                    ref = r_view
+                elif (r_view - ref).abs().max().item() > 1e-2:      # view not stable across calls
+                    ok = False; break
+            safe = ok
+    except Exception:
+        safe = False   # anything unexpected -> keep the safe contiguous copy
+
+    _TRANSPOSED_VIEW_GROUPED_MM_SAFE = safe
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils.py:366]: Not unused: it is read in _transposed_view_grouped_mm_is_safe as the None-sentinel cache check before the probe runs.

File: tests/test_moe_recompute_gc_adaptive.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            assert in_gradient_checkpoint_recompute() is True
+        assert in_gradient_checkpoint_recompute() is True    # restored to outer
+    assert in_gradient_checkpoint_recompute() is False       # restored to base
+
+    with pytest.raises(ValueError):
+        with _gradient_checkpoint_recompute_marker():
+            assert in_gradient_checkpoint_recompute() is True
+            raise ValueError("boom")
+    # The finally clause must clear the flag even when the body raised.
+    assert in_gradient_checkpoint_recompute() is False
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_moe_recompute_gc_adaptive.py:43]: Reachable: this runs after the pytest.raises block returns, asserting the marker was cleared by the context manager finally even when the body raised.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Changed the default for bnb 4-bit grouped MoE under gradient checkpointing: it now recomputes instead of pinning (commit f0e0974f).

Rationale: the adaptive policy pins the dequantized base inside a GC recompute pass because the pin is momentary. But for a 4-bit MoE that momentary pin still materializes the full bf16 expert stack the 4-bit storage exists to avoid, which is several GiB per layer on large MoEs, exactly when a 4-bit + gradient-checkpointing run is most memory-constrained. Pin and recompute are dX-identical for the frozen base, so this only trades a little speed for memory.

Scope: keyed on the source being a frozen Params4bit (via _source_pins_large_dequant). Plain bf16/fp16/fp32 bases pin their own storage with no extra dequant, so they keep the speed-oriented adaptive policy unchanged. UNSLOTH_MOE_RECOMPUTE=0 still forces the pin for memory-rich runs; =1 still forces recompute.

Measured tradeoff (from the earlier bench on unsloth/gpt-oss-20b-unsloth-bnb-4bit, GC on): the old pin default was 16.88 GB / 6718 tok/s; the new recompute default is 16.40 GB / 6439 tok/s. So on this model it is about -0.48 GB for about -4% throughput, with the memory saving growing with expert size and preventing the pin-path OOM on larger 4-bit MoEs. The dispatch gate and the per-source decision in forward_native_grouped_mm were updated together so they stay consistent, with unit tests over the 4-bit vs dense split.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
### Final results

Setup: `unsloth/gpt-oss-20b-unsloth-bnb-4bit`, attention-only LoRA, bf16, bsz 2 x grad-accum 3, seq 2048, 61 steps, single B200. The grouped 4-bit path engaged in every run. Peak GB is `max_memory_allocated` across steps; run-to-run noise floor 0.5 percent. Pin and recompute are dX-identical for the frozen base, so this trades speed for memory only (loss spread within the bf16 noise floor).

Gradient checkpointing ON (the memory-constrained regime this PR targets):

| 4-bit expert base | Peak GB | Step ms | tok/s |
|-------------------|---------|---------|-------|
| pin the bf16 dequant (`UNSLOTH_MOE_RECOMPUTE=0`) | 16.88 | 1825 | 6734 |
| recompute in backward (new default) | 16.40 | 1908 | 6439 |

Gradient checkpointing OFF:

| 4-bit expert base | Peak GB | Step ms | tok/s |
|-------------------|---------|---------|-------|
| pin (`=0`) | 17.23 | 2477 | 4961 |
| recompute (default) | 16.74 | 2686 | 4574 |

Takeaway: for a bnb 4-bit MoE the default now recomputes the expert stack in backward instead of pinning its full bf16 dequant, saving 0.48 GB on gpt-oss-20b under gradient checkpointing (about 4 percent throughput), with the saving growing with expert size and preventing the pin-path OOM on larger 4-bit MoEs. Plain bf16/fp16/fp32 bases keep the speed-oriented adaptive policy (pin under a GC recompute pass, recompute otherwise). `UNSLOTH_MOE_RECOMPUTE=0` forces the pin for memory-rich runs; `=1` forces recompute.
